book/src/dev/rfcs/xxxx-block-subsidy.md

```diff
@@ -0,0 +1,296 @@
+
+### 2 - Funding stream:
+
+*[Canopy onward]  The coinbase transaction at `height` **MUST**  contain at least one output per funding stream `fs` active at
+height, that pays `fs.Value(height)` zatoshi in the prescribed way to the stream’s recipient address represented by `fs.AddressList` of `fs.AddressIndex(height)`.* https://zips.z.cash/protocol/protocol.pdf#fundingstreams
+
+We make use of the funding streams functions here, similar to founders reward . We get the amount of the reward using `funding_stream(height, network)` and then the address with `funding_stream_address(height, network)`. Next we get a list of outputs that match the amount with the utility function `find_output_with_amount()`. Finally with this list, we check if any of the output scripts matches the address we have computed.
+
+### 3 - Miner subsidy:
+
+*The total amount of transparent outputs from a coinbase transaction, minus the amount of the `valueBalance` field if present, **MUST NOT** be greater than the amount of miner subsidy plus the total amount of transaction fees paid by transactions in this block.* https://zips.z.cash/protocol/canopy.pdf#txnencodingandconsensus
+
+So the rule is the following before Canopy:
+
+`coinbase_sum_outputs() <= miner_subsidy() + founders_reward() + transaction_fees()` 
+
+and after Canopy:
+
+`coinbase_sum_outputs() <= miner_subsidy() + funding_stream() + transaction_fees()`
+
+### 4 - Shielded coinbase:
+
+*[Pre-Heartwood] A coinbase transaction MUST NOT have any JoinSplit descriptions, Spend descriptions, or Output descriptions.*
+
+*[Heartwood onward] A coinbase transaction MUST NOT have any JoinSplit descriptions or Spend descriptions.*
+
+*[Heartwood onward] The consensus rules applied to `valueBalance`, `vShieldedOutput`, and `bindingSig` in non-coinbase transactions MUST also be applied to coinbase transactions.*
+
+https://zips.z.cash/zip-0213#specification
+
+This rules are implemented inside `shielded_coinbase()`.
+
+### Validation errors
+
+A `SubsidyError` type will be created to handle validation of the above consensus rules:
+
+```
+pub enum SubsidyError {
+    #[error("not a coinbase transaction")]
+    NoCoinbase,
+
+    #[error("founders reward amount not found")]
+    FoundersRewardAmountNotFound,
+    
+    #[error("founders reward address not found")]
+    FoundersRewardAddressNotFound,
+    
+    #[error("founding stream amount not found")]
+    FundingStreamAmountNotFound,
+    
+    #[error("founding stream address not found")]
+    FundingStreamAddressNotFound,
+    
+    #[error("the sum of outputs is greater than calculated subsidies and transaction fees.")]
+    MinerSubsidyRuleBroken,
+    
+    #[error("invalid shielded descriptions found.")]
+    ShieldedDescriptionsInvalid,
+
+    #[error("broken rule in shielded transaction inside coinbase.")]
+    ShieldedRuleBroken,
+}
+```
```

Comment:
@yaahc can you review the error names and messages for block subsidy?

Reply:
The error types names look great, fwiw I ended up referring to `std::io::ErrorKind` for a reference on how to structure the errors and message, https://doc.rust-lang.org/stable/src/std/io/error.rs.html#90-178